qutip/solver/mcsolve.py

```diff
@@ -0,0 +1,520 @@
+        Seed for the random number generator. It can be a single seed used to
+        spawn seeds for each trajectory or a list of seed, one for each
+        trajectory. Seeds are saved in the result and can be reused with::
+            seeds=prev_result.seeds
+    """
+    name = "mcsolve"
+    resultclass = McResult
+    solver_options = {
+        "progress_bar": "text",
+        "progress_kwargs": {"chunk_size": 10},
+        "store_final_state": False,
+        "store_states": None,
+        "keep_runs_results": False,
+        "method": "adams",
+        "map": "serial",
+        "job_timeout": None,
+        "num_cpus": None,
+        "bitgenerator": None,
+        "mc_corr_eps": 1e-10,
+        "norm_steps": 5,
+        "norm_t_tol": 1e-6,
+        "norm_tol": 1e-4,
+    }
+
+    def __init__(self, H, c_ops, *, options=None):
+        _time_start = time()
+
+        if isinstance(c_ops, (Qobj, QobjEvo)):
+            c_ops = [c_ops]
+        c_ops = [QobjEvo(c_op) for c_op in c_ops]
+
+        if H.issuper:
+            self._c_ops = [
+                spre(c_op) * spost(c_op.dag()) if c_op.isoper else c_op
+                for c_op in c_ops
+            ]
+            self._n_ops = self._c_ops
+            rhs = QobjEvo(H)
+            for c_op in c_ops:
+                cdc = c_op.dag() @ c_op
+                rhs -= 0.5 * (spre(cdc) + spost(cdc))
+        else:
+            self._c_ops = c_ops
+            self._n_ops = [c_op.dag() * c_op for c_op in c_ops]
+            rhs = -1j * QobjEvo(H)
+            for n_op in self._n_ops:
+                rhs -= 0.5 * n_op
+
+        self._num_collapse = len(self._c_ops)
+
+        super().__init__(rhs, options=options)
+
+    def _restore_state(self, data, *, copy=True):
+        """
+        Retore the Qobj state from its data.
+        """
+        if self._state_metadata['dims'] == self.rhs.dims[1]:
+            state = Qobj(unstack_columns(data),
+                         **self._state_metadata, copy=False)
+        else:
+            state = Qobj(data, **self._state_metadata, copy=copy)
+
+        return state
+
+    def _initialize_stats(self):
+        stats = super()._initialize_stats()
+        stats.update({
+            "method": self.options["method"],
+            "solver": "Master Equation Evolution",
+            "num_collapse": self._num_collapse,
+        })
+        return stats
+
+    def _argument(self, args):
+        self._integrator.arguments(args)
+        self.rhs.arguments(args)
+        for c_op in self._c_ops:
+            c_op.arguments(args)
+        for n_op in self._n_ops:
+            n_op.arguments(args)
+
+    def _run_one_traj(self, seed, state, tlist, e_ops):
+        """
+        Run one trajectory and return the result.
+        """
+        result = Result(e_ops, {**self.options, "normalize_output": False})
+        generator = self._get_generator(seed)
+        self._integrator.set_state(tlist[0], state, generator)
```

Comment:
Setting the integrator state and running the integrator is non-reentrant. That's safe if the parallelization forks the process, but not if it creates a thread. If would be good to solve this if we can, or just to leave a comment here if it looks like it will create too much complexity or overhead to use a separate integrator for each run of this function.

Reply:
Since we don't have a thread based map ,and probably never will, I am only leaving a comment.

Some of the integrators can be slow to create, but we could just add a `copy` method to the class. However, the scipy integrator themselves are non-reentrant, while we mitigated it a little in #1734, it would not survive multi-threading. The integrator we made should be fine.